pybop/optimisers/pints_optimisers.py

```diff
@@ -1,8 +1,428 @@
+        Enable or disable parallel evaluation.
+        Credit: PINTS
+
+        Parameters
+        ----------
+        parallel : bool or int, optional
+            If True, use as many worker processes as there are CPU cores. If an integer, use that many workers.
+            If False or 0, disable parallelism (default: False).
+        """
+        if parallel is True:
+            self._parallel = True
+            self._n_workers = pints.ParallelEvaluator.cpu_count()
+        elif parallel >= 1:
+            self._parallel = True
+            self._n_workers = int(parallel)
+        else:
+            self._parallel = False
+            self._n_workers = 1
+
+    def set_min_iterations(self, iterations=2):
+        """
+        Set the minimum number of iterations as a stopping criterion.
+
+        Parameters
+        ----------
+        iterations : int, optional
+            The minimum number of iterations to run (default: 2).
+            Set to `None` to remove this stopping criterion.
+        """
+        if iterations is not None:
+            iterations = int(iterations)
+            if iterations < 0:
+                raise ValueError("Minimum number of iterations cannot be negative.")
+        self._min_iterations = iterations
+
+    def set_max_unchanged_iterations(self, iterations=15, threshold=1e-5):
+        """
+        Set the maximum number of iterations without significant change as a stopping criterion.
+        Credit: PINTS
+
+        Parameters
+        ----------
+        iterations : int, optional
+            The maximum number of unchanged iterations to run (default: 15).
+            Set to `None` to remove this stopping criterion.
+        threshold : float, optional
+            The minimum significant change in the objective function value that resets the
+            unchanged iteration counter (default: 1e-5).
+        """
+        if iterations is not None:
+            iterations = int(iterations)
+            if iterations < 0:
+                raise ValueError("Maximum number of iterations cannot be negative.")
+
+        threshold = float(threshold)
+        if threshold < 0:
+            raise ValueError("Minimum significant change cannot be negative.")
+
+        self._unchanged_max_iterations = iterations
+        self._unchanged_threshold = threshold
+
+    def set_max_evaluations(self, evaluations=None):
+        """
+        Set a maximum number of evaluations stopping criterion.
+        Credit: PINTS
+
+        Parameters
+        ----------
+        evaluations : int, optional
+            The maximum number of evaluations after which to stop the optimisation
+            (default: None).
+        """
+        if evaluations is not None:
+            evaluations = int(evaluations)
+            if evaluations < 0:
+                raise ValueError("Maximum number of evaluations cannot be negative.")
+        self._max_evaluations = evaluations
+
+
+class GradientDescent(BasePintsOptimiser):
```

Comment:
I think it would be good split the pints' optimisers into a separate file to reduce the length of this file.